provider/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinBuildScriptCompiler.kt

```diff
@@ -79,73 +80,72 @@ class KotlinBuildScriptCompiler(
-            ignoringErrors { executeScriptBodyOn(target) }
+        asKotlinScript {
+            ignoringErrors { executeBuildscriptBlock() }
+            ignoringErrors { prepareTargetClassLoaderScope() }
+            ignoringErrors { executeScriptBody() }
         }
 
     private
     fun compileTopLevelScript() =
-        asKotlinScript { target ->
+        asKotlinScript {
             withUnexpectedBlockHandling {
-                executeBuildscriptBlockOn(target)
-                prepareAndExecuteScriptBodyOn(target)
+                executeBuildscriptBlock()
+                prepareAndExecuteScriptBody()
             }
         }
 
     private
     fun compileScriptPlugin() =
-        asKotlinScript { target ->
+        asKotlinScript {
             withUnexpectedBlockHandling {
-                prepareAndExecuteScriptBodyOn(target)
+                prepareAndExecuteScriptBody()
             }
         }
 
     private
-    fun asKotlinScript(script: (KotlinScriptTarget<*>) -> Unit): KotlinScript = { target ->
-        val scriptTarget = kotlinScriptTargetFor(target)
+    fun asKotlinScript(script: () -> Unit): KotlinScript = {
         scriptTarget.prepare()
-        script(scriptTarget)
+        script()
     }
 
     private
-    fun prepareAndExecuteScriptBodyOn(target: KotlinScriptTarget<*>) {
-        prepareTargetClassLoaderScopeOf(target)
-        executeScriptBodyOn(target)
+    fun prepareAndExecuteScriptBody() {
+        prepareTargetClassLoaderScope()
+        executeScriptBody()
     }
 
     private
-    fun executeScriptBodyOn(target: KotlinScriptTarget<*>) {
-        val accessorsClassPath = accessorsClassPathFor(target)
-        val compiledScript = compileScriptFileFor(target, compilationClassPath + accessorsClassPath)
+    fun executeScriptBody() {
+        val accessorsClassPath = accessorsClassPath()
+        val compiledScript = compileScriptFileFor(compilationClassPath + accessorsClassPath)
         val scriptScope = scriptClassLoaderScopeWith(accessorsClassPath)
-        executeCompiledScript(compiledScript, scriptScope, target)
+        executeCompiledScript(compiledScript, scriptScope)
     }
 
     private
-    fun accessorsClassPathFor(target: KotlinScriptTarget<*>): ClassPath =
-        target.takeIf { topLevelScript }?.accessorsClassPathFor(compilationClassPath)?.bin
+    fun accessorsClassPath(): ClassPath =
+        scriptTarget.takeIf { topLevelScript }?.accessorsClassPathFor(compilationClassPath)?.bin
             ?: ClassPath.EMPTY
 
     private
     fun scriptClassLoaderScopeWith(accessorsClassPath: ClassPath) =
         targetScope.createChild("script").apply { local(accessorsClassPath) }
 
     private
-    fun executeBuildscriptBlockOn(target: KotlinScriptTarget<*>) {
+    fun executeBuildscriptBlock() {
         setupEmbeddedKotlinForBuildscript()
-        if (target.supportsBuildscriptBlock) {
+        if (scriptTarget.supportsBuildscriptBlock) {
```

Comment:
:o: I wonder if this property is really necessary or if this thing would be better expressed by mapping over the nullable `buildscriptBlockTemplate`.

```kotlin
scriptTarget.buildscriptBlockTemplate?.let { template ->
    extractBuildscriptBlockFrom(script)?.let { buildscriptRange ->
      executeBuildscriptBlockFrom(buildscriptRange, template)
    }
}
```

Passing the resolved template down would avoid the double bang at least. WDYT?

Reply:
I made the change for both `buildscript` & `plugins`, for symmetry.